tpp-app/src/services/CalculationService.js

```diff
@@ -0,0 +1,167 @@
+    return todayPeriod && yesterdayNoPeriod && twoDaysEarlierNoPeriod;
+};
+
+/**
+ * Returns True if today is the start of a period, False otherwise i.e. matches period pattern of _ _ X
+ * @return {boolean} Returns a boolean indicating if period has started today
+ */
+export const isPeriodStarting = async () => {
+    const lastThreeDaysSymptoms = await GETLastThreeDaysSymptoms();
+    const twoDaysEarlierSymptoms = lastThreeDaysSymptoms[0];
+    const yesterdaySymptoms = lastThreeDaysSymptoms[1];
+    const todaySymptoms = lastThreeDaysSymptoms[2];
+
+    const todayPeriod = (todaySymptoms.flow != null && todaySymptoms.flow !== FLOW_LEVEL.NONE);
+    const yesterdayNoPeriod = (yesterdaySymptoms.flow == null && yesterdaySymptoms.flow === FLOW_LEVEL.NONE);
+    const twoDaysEarlierNoPeriod = (twoDaysEarlierSymptoms.flow == null && twoDaysEarlierSymptoms.flow === FLOW_LEVEL.NONE);
+    return todayPeriod && yesterdayNoPeriod && twoDaysEarlierNoPeriod;
+}
+
+/**
+ * Calculates the average period length of the user.
+ */
+export const calculateAveragePeriodLength = async () => new Promise( async (resolve, reject) => {
+    // Only calculate Average Period Length once period is over
+    if (isPeriodOver()) {
+        // Record all period intervals
+        let completeHistory = [];
+
+        // Get all Years
+        GETStoredYears()
+            .then((years) => {
+                years.forEach(async (year) => {
+                    completeHistory.push.apply(await CycleService.GETCycleHistoryByYear(year));
+                });
+            })
+            .catch((e) => {
+                console.log(`calculateAveragePeriodLength error: ${JSON.stringify(error)}`);
+                reject();
+            });
+
+        // Use reduce to find average
+        const averagePeriodLength = completeHistory.reduce(function (sum, interval) {
+            return sum + interval.periodDays;
+        }, 0) / completeHistory.length;
+
+        await AsyncStorage.Set(
+            KEYS.AVERAGE_PERIOD_LENGTH, JSON.stringify(averagePeriodLength)
+        ).then(() => {
+            console.log(`Recalculated ${KEYS.AVERAGE_PERIOD_LENGTH} as ${averagePeriodLength}`)
+            resolve();
+        }).catch((error) => {
+            console.log(`calculateAveragePeriodLength error: ${JSON.stringify(error)}`);
+            reject();
+        });
+    }
+
+    // Else don't calculate, do nothing
+});
+
+/**
+ * Calculates the average cycle length of the user.
+ */
+export const calculateAverageCycleLength = async () => new Promise( async (resolve, reject) => {
+    // Only calculate Average Cycle Length once period is starting and cycle is over
+    if (isPeriodStarting()) {
+        let completeHistory = [];
+
+        // Get all Years
+        GETStoredYears()
+            .then((years) => {
+                years.forEach(async (year) => {
+                    completeHistory.push.apply(await CycleService.GETCycleHistoryByYear(year));
```

Comment:
why do u need to use apply here? just curious

Reply:
.push.apply() will flatten the list returned from GETCycleHistoryByYear before adding it to completeHistory because I  want to use 
```
completeHistory = [{"start": "2022-12-20", "periodDays": 5}, {"start": "2022-01-10", "periodDays": 5}, {"start": "2022-02-05", "periodDays": 4} ...]
```
and not nested like 
```
completeHistory = [[{"start": "2022-12-20", "periodDays": 5}], [{"start": "2022-12-20", "periodDays": 5}, {"start": "2022-02-05", "periodDays": 4}], ...]
```
in .map() and .reduce()

Reply:
https://stackoverflow.com/questions/4007744/how-to-do-a-flat-push-in-javascript